californium-core/src/main/java/org/eclipse/californium/core/network/stack/ReliabilityLayer.java

```diff
@@ -364,6 +365,15 @@ public void run() {
 					LOGGER.log(Level.FINE, "Timeout: retransmission limit reached, exchange failed, message: {0}", message);
 					exchange.setTimedOut();
 					message.setTimedOut(true);
+					OptionSet options = message.getOptions();
+					boolean isBlockwise = options.hasBlock1() || options.hasBlock2();
+					if (message instanceof Request && isBlockwise) {
+					    Request request = exchange.getRequest();
+					    if (request != null && request.getMID() != message.getMID()) {
+					        // Notifying the original request about timeout
+					        request.setTimedOut(true);
+					    }
```

Comment:
It is hard for me to figure out what could be the intent to make a listener for an intermediate request, but your last point looks very reasonable, I think such behavior seems as more expected by a user.